File: app/src/main/java/com/meiji/daily/module/useradd/UserAddViewModel.java
package com.meiji.daily.module.useradd;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;
import android.content.SharedPreferences;
import android.support.annotation.NonNull;

import com.meiji.daily.Constant;
import com.meiji.daily.bean.ZhuanlanBean;
import com.meiji.daily.data.local.AppDatabase;
import com.meiji.daily.data.remote.IApi;
import com.meiji.daily.util.ErrorAction;
import com.meiji.daily.util.RxBusHelper;

import java.util.List;

import io.reactivex.Flowable;
import io.reactivex.Single;
import io.reactivex.SingleEmitter;
import io.reactivex.SingleOnSubscribe;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;
import retrofit2.Retrofit;

/**
 * Created by Meiji on 2017/12/4.
 */

public class UserAddViewModel extends AndroidViewModel implements SharedPreferences.OnSharedPreferenceChangeListener {

    private final AppDatabase mAppDatabase;
    private final Retrofit mRetrofit;
    private final RxBusHelper mRxBusHelper;
    private final CompositeDisposable mDisposable;

    private Flowable<Boolean> mRxBus;
    private MutableLiveData<Boolean> mIsLoading;
    private MutableLiveData<Boolean> mIsRefreshUI;
    private MutableLiveData<Boolean> mIsAddSuccess;
    private MutableLiveData<List<ZhuanlanBean>> mList;

    {
        mIsLoading = new MutableLiveData<>();
        mIsRefreshUI = new MutableLiveData<>();
        mIsAddSuccess = new MutableLiveData<>();
        mList = new MutableLiveData<>();
        mDisposable = new CompositeDisposable();

        mIsLoading.setValue(true);
        mIsRefreshUI.setValue(true);
    }

    UserAddViewModel(Application application, AppDatabase appDatabase,
                     Retrofit retrofit, RxBusHelper rxBusHelper) {
        super(application);
        mAppDatabase = appDatabase;
        mRetrofit = retrofit;
        mRxBusHelper = rxBusHelper;

        handleData();
        subscribeTheme();
    }

    public MutableLiveData<List<ZhuanlanBean>> getList() {
        return mList;
    }

    MutableLiveData<Boolean> isRefreshUI() {
        return mIsRefreshUI;
    }

    MutableLiveData<Boolean> isLoading() {
        return mIsLoading;
    }

    MutableLiveData<Boolean> isAddResult() {
        return mIsAddSuccess;
    }

    void handleData() {
        mIsLoading.setValue(true);

        Disposable subscribe = mAppDatabase.ZhuanlanNewDao().query(Constant.TYPE_USERADD)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Consumer<List<ZhuanlanBean>>() {
                    @Override
                    public void accept(List<ZhuanlanBean> list) throws Exception {
                        mList.setValue(list);
                    }
                }, ErrorAction.error());
        mDisposable.add(subscribe);
        mIsLoading.setValue(false);
    }

    void addItem(final String input) {
        mIsLoading.setValue(true);

        Disposable subscribe = mRetrofit.create(IApi.class).getZhuanlanBean(input)
                .subscribeOn(Schedulers.io())
                .doOnSuccess(new Consumer<ZhuanlanBean>() {
                    @Override
                    public void accept(ZhuanlanBean bean) throws Exception {
                        if (bean != null) {
                            bean.setType(Constant.TYPE_USERADD);
                            mAppDatabase.ZhuanlanNewDao().insert(bean);
                        }
                    }
                })
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Consumer<ZhuanlanBean>() {
                    @Override
                    public void accept(@io.reactivex.annotations.NonNull ZhuanlanBean bean) throws Exception {
                        mIsAddSuccess.setValue(true);
                        handleData();
                    }
                }, new ErrorAction() {
                    @Override
                    public void doAction() {
                        mIsAddSuccess.setValue(false);
                        mIsLoading.setValue(false);
                    }
                }.action());
        mDisposable.add(subscribe);
    }

    private void subscribeTheme() {
        mRxBus = mRxBusHelper.register(Constant.RxBusEvent.REFRESHUI);
        Disposable subscribe = mRxBus.subscribe(new Consumer<Boolean>() {
            @Override
            public void accept(Boolean aBoolean) throws Exception {
                mIsRefreshUI.setValue(mIsRefreshUI.getValue() != null && !mIsRefreshUI.getValue());
            }
        }, ErrorAction.error());
        mDisposable.add(subscribe);
    }

    void deleteItem(final ZhuanlanBean bean) {
        Disposable subscribe = Single.create(new SingleOnSubscribe<Object>() {
            @Override
            public void subscribe(SingleEmitter<Object> e) throws Exception {
                mAppDatabase.ZhuanlanNewDao().delete(bean.getSlug());
            }
        }).subscribeOn(Schedulers.io()).subscribe();
        mDisposable.add(subscribe);
    }

    @Override
    protected void onCleared() {
        mRxBusHelper.unregister(Constant.RxBusEvent.REFRESHUI, mRxBus);
        mDisposable.clear();
        super.onCleared();
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {

    }

    public static class Factory extends ViewModelProvider.NewInstanceFactory {

        private final Application mApplication;
        private final AppDatabase mAppDatabase;
        private final Retrofit mRetrofit;
        private final RxBusHelper mRxBusHelper;

        Factory(Application application, AppDatabase appDatabase,
                Retrofit retrofit, RxBusHelper rxBusHelper) {
            mApplication = application;
            mAppDatabase = appDatabase;
            mRetrofit = retrofit;
            mRxBusHelper = rxBusHelper;
        }

        @NonNull
        @Override
        public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
            return (T) new UserAddViewModel(mApplication, mAppDatabase, mRetrofit, mRxBusHelper);
        }
    }
}

File: app/src/main/java/com/meiji/daily/module/useradd/UserAddViewModel.kt
package com.meiji.daily.module.useradd

import android.app.Application
import android.arch.lifecycle.AndroidViewModel
import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel
import android.arch.lifecycle.ViewModelProvider
import com.meiji.daily.Constant
import com.meiji.daily.bean.ZhuanlanBean
import com.meiji.daily.data.local.AppDatabase
import com.meiji.daily.data.remote.IApi
import com.meiji.daily.io
import com.meiji.daily.mainThread
import com.meiji.daily.util.ErrorAction
import com.meiji.daily.util.RxBusHelper
import io.reactivex.Flowable
import io.reactivex.Single
import io.reactivex.SingleOnSubscribe
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.functions.Consumer
import retrofit2.Retrofit

/**
 * Created by Meiji on 2017/12/4.
 */

class UserAddViewModel
constructor(application: Application,
            private val mAppDatabase: AppDatabase,
            private val mRetrofit: Retrofit,
            private val mRxBusHelper: RxBusHelper)
    : AndroidViewModel(application) {

    private val mDisposable: CompositeDisposable

    private lateinit var mRxBus: Flowable<Any>
    var isLoading: MutableLiveData<Boolean>
        private set
    var isRefreshUI: MutableLiveData<Boolean>
        private set
    var isAddResult: MutableLiveData<Boolean>
        private set
    var mList: MutableLiveData<MutableList<ZhuanlanBean>>
        private set

    init {
        isLoading = MutableLiveData()
        isRefreshUI = MutableLiveData()
        isAddResult = MutableLiveData()
        mList = MutableLiveData()
        mDisposable = CompositeDisposable()

        isLoading.value = true
        isRefreshUI.value = true
    }

    init {
        handleData()
        subscribeTheme()
    }

    fun handleData() {
        isLoading.value = true

        mAppDatabase.zhuanlanDao().query(Constant.TYPE_USERADD)
                .subscribeOn(io)
                .observeOn(mainThread)
                .subscribe(Consumer<MutableList<ZhuanlanBean>> { list ->
                    mList.value = list
                }, ErrorAction.error()).let { mDisposable.add(it) }
        isLoading.value = false
    }

    internal fun addItem(input: String) {
        isLoading.value = true

        mRetrofit.create(IApi::class.java).getZhuanlanBean(input)
                .doOnSuccess { bean ->
                    bean?.let {
                        it.type = Constant.TYPE_USERADD
                        mAppDatabase.zhuanlanDao().insert(it)
                    }
                }
                .subscribeOn(io)
                .observeOn(mainThread)
                .subscribe(Consumer {
                    isAddResult.value = true
                    handleData()
                }, object : ErrorAction() {
                    override fun doAction() {
                        isAddResult.value = false
                        isLoading.value = false
                    }
                }.action()).let { mDisposable.add(it) }
    }

    private fun subscribeTheme() {
        mRxBus = mRxBusHelper.register(Constant.RxBusEvent.REFRESHUI)
        mRxBus.subscribe(Consumer {
            isRefreshUI.setValue(!(isRefreshUI.value)!!)
        }, ErrorAction.error()).let { mDisposable.add(it) }
    }

    internal fun deleteItem(bean: ZhuanlanBean) {
        Single.create(SingleOnSubscribe<Any> {
            mAppDatabase.zhuanlanDao().delete(bean.slug)
        }).subscribeOn(io)
                .subscribe().let { mDisposable.add(it) }

    }

    override fun onCleared() {
        mRxBusHelper.unregister(Constant.RxBusEvent.REFRESHUI, mRxBus)
        mDisposable.clear()
        super.onCleared()
    }

    class Factory internal constructor(private val mApplication: Application,
                                       private val mAppDatabase: AppDatabase,
                                       private val mRetrofit: Retrofit,
                                       private val mRxBusHelper: RxBusHelper)
        : ViewModelProvider.NewInstanceFactory() {

        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            return UserAddViewModel(mApplication, mAppDatabase, mRetrofit, mRxBusHelper) as T
        }
    }
}

